### PR TITLE
[Torchax] Support Running AWQ models

### PR DESCRIPTION
**Pull request by kyuyeunk**
# Description

- Following PR adds support for running loading & running AWQ models in torchax
  - Please note that this requires a fix in vLLM repo to be submitted as well: https://github.com/vllm-project/vllm/pull/23902
- Using model [Qwen/Qwen2.5-32B-Instruct-AWQ](https://huggingface.co/Qwen/Qwen2.5-32B-Instruct-AWQ), I have verified that it'ss running correctly.
	- Numeric validation
		- Command `TPU_BACKEND_TYPE=jax MODEL_IMPL_TYPE=vllm python3 examples/offline_inference.py --model=Qwen/Qwen2.5-32B-Instruct-AWQ --tensor_parallel_size=8 --task=generate --max_model_len=1024 --download_dir=/mnt/disks/persist`
		- Result: http://gpaste/6205721003425792
	- Performance validation (degradation over baseline is expected)
		- Baseline ([Qwen/Qwen2.5-32B-Instruct](https://huggingface.co/Qwen/Qwen2.5-32B-Instruct)): 12.61 reqs/s
		- AWQ ([Qwen/Qwen2.5-32B-Instruct-AWQ](https://huggingface.co/Qwen/Qwen2.5-32B-Instruct-AWQ)): 11.34 reqs/s
- Limitation / Future work
	- AWQ uses w4a16 subchannel quantization, and for initial implementation, I have utilized XLA instead of kernel. However, there is known performance issue with subchannel quantization in XLA. Therefore, we need to implement w4a16 subchannel support into the kernel for optimal performance.

# Tests

```
pytest -v tests/models/vllm/layers/test_awq.py
```

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_commons/models/vllm/quantization/awq.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+        group_size = self.quant_config.group_size
+        # Reshape so that each qweight[i] were quantized with same scales[i].
+        qweight = qweight.reshape((-1, group_size, layer.output_size))
+        qweight = torch_to_jax_param(
+            qweight,
+            NamedSharding(self.jax_config.mesh,
+                          self.jax_config.weight_sharding),
+            self.jax_config.output_sizes, self.jax_config.n_shards,
+            self.jax_config.fuse_matmuls, 2, jnp.uint4)
```
Comment (hfan) [inline tpu_commons/models/vllm/quantization/awq.py]: nit: it helps the reader if the dim arg is specified using the keyword arg syntax, i.e. `dim=2`

File: tests/models/vllm/layers/test_awq.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+    if b is not None:
+        out += b
+    return out.to(x.dtype)
+
+
+def pack_awq_weight_into_int32(weight: torch.Tensor):
+    # AWQ packs 8 uint4 into 32-bits in this order.
+    awq_order = (0, 2, 4, 6, 1, 3, 5, 7)
+
+    orig_shape = weight.shape
```
Comment (vanbasten23) [inline tests/models/vllm/layers/test_awq.py:72]: add this test "pytest -v tests/models/vllm/layers/test_awq.py" to the CI?

File: tpu_commons/models/vllm/quantization/awq.py (hunk 1 above)
Comment (kyuyeunk) [reply tpu_commons/models/vllm/quantization/awq.py]: Done.

File: tests/models/vllm/layers/test_awq.py (hunk 2 above)
Comment (kyuyeunk) [reply tests/models/vllm/layers/test_awq.py:72]: This gets run in Jax unit test: https://github.com/vllm-project/tpu_commons/blob/4470371755e316bf7c3a5e5940133923d7ebc325/.buildkite/pipeline_jax.yml#L76-L88

In a (failed...) CI, I have confirmed that is the case: https://buildkite.com/tpu-commons/tpu-commons-ci/builds/2546#01990bbf-4525-4e7c-ad33-667076f9482d

Is it a better idea to add this in torch ci?: https://github.com/vllm-project/tpu_commons/blob/main/.buildkite/pipeline_torch.yml

Comment (vanbasten23) [reply tests/models/vllm/layers/test_awq.py:72]: I see. Thanks. If it's framework agnostic, then it's fine to leave it as is.
